### DISCUSSION TITLE
JSON config isn't showing default value when the params value is set to a dictionary

### DISCUSSION THREAD

**Discussion by rustikk** [apache/airflow#20785] [Q&A]
### Apache Airflow version

2.2.3 (latest released)

### What happened

When triggering a dag with the JSON config with the params value set to a dictionary with type="object" the default value is not displaying. The schema validation is still working, if you pass a value that isn't a dictionary an error will display in the UI. 

### What you expected to happen

I expected the default value to show in the JSON config.

### How to reproduce

```
from datetime import datetime, timedelta

from airflow import DAG
from airflow.models.param import Param
from airflow.operators.python import PythonOperator

# unpausing causes this dag to run once, because there's only been
# one complete interval of 30 years since the epoch.
thirty_years = 1577862000  # seconds

with DAG(
    "params_dictionaries",
    start_date=datetime(1970, 1, 1),
    schedule_interval=timedelta(seconds=thirty_years),
    catchup=True,
    params={"x": Param({"a": {3, 4, 5}, "b": {1, 2, 3}}, type="object")},
    render_template_as_native_obj=True,
    ),
) as dag:

    def fail_if_invalid(val):
        print(val)
        # assert type(val) == dict

    PythonOperator(
        task_id="ref_param",
        python_callable=fail_if_invalid,
        op_args=["{{ params['x'] }}"]
    )
```

### Operating System

Docker (debian:buster)

### Versions of Apache Airflow Providers

2.2.3, 2.2.2, 2.2.1 gives an import error

### Deployment

Astronomer

### Deployment details

FROM quay.io/astronomer/ap-airflow-dev:2.2.3-onbuild
#FROM quay.io/astronomer/ap-airflow:2.2.2-buster-onbuild
#FROM quay.io/astronomer/ap-airflow-dev:2.2.1-buster-onbuild

### Anything else

Problem occurs everytime.

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by potiuk** [comment]
Converted it into discussion as it is not clear what is wrong. Could you please elaborate (including screenshots, error messages) what do you mean by "value showing in JSON config"?  Including some example of showing "good" behaviour with non-dict params and bad behaviour with dict params? It's very likely misunderstanding on how params work so I also encorage you to take a look there: https://airflow.apache.org/docs/apache-airflow/stable/concepts/params.html?highlight=params

@msumit  -  is this the right "params" definition?

**Reply by msumit** [reply]
@potiuk the use case seems to be valid & the json-schema also validates it correctly. @rustikk does the UI work differently when you don't pass a `Param` object & simply pass a dictionary or use an Airflow version that is predated params concept?

**Comment by MatrixManAtYrService** [comment]
I think the missing link here is that `{"a": {3, 4, 5}, "b": {1, 2, 3}}` is a valid python object, but not a valid JSON object (JSON doesn't have sets).

Typically, when you "trigger DAG w/ config", you see a screen like this:
![image](https://user-images.githubusercontent.com/5834582/149260140-7367fadc-e50f-4919-bbfa-9cd3b92ddba4.png)

Where the text area is populated with the JSON version of whatever the default argument is.  But since there is no valid json representation of the default, that area is blank, as @rustikk has pointed out.

This only comes up when accepting input from the user.  The scheduler will happily create dags that use the problematic default value.  If you unpause this dag, the assertion will pass:

```python3
default_obj = {"foo": {"bar", "baz"}}

@task
def usedict(**context):
    referenced_obj = context["params"]["x"]
    assert "baz" in referenced_obj["foo"]

with DAG(
    "params_dictionaries",
    start_date=datetime(1970, 1, 1),
    schedule_interval=timedelta(seconds=1577862000),
    catchup=True,
    params={"x": Param(default_obj, type="object")},
    render_template_as_native_obj=True,
) as dag:
    usedict()
```

If you want to be able to edit the defaults that appear in that text area, you'll be able to do so only if you start with a json serializable default.  For instance, making this change fixes it:
```
# default_obj = {"foo": {"bar", "baz"}}  # dict with set
default_obj = {"foo": {"bar": "baz"}}  # dict with dict
```

I think we should check if the Param default is json serializable and raise a TypeError at parse-time if it isn't:
> Parameter validation uses JsonSchema, but Airflow could not represent the default for parameter "x" as json:
> TypeError: Object of type set is not JSON serializable

**Reply by MatrixManAtYrService** [reply]
Related: https://github.com/apache/airflow/pull/20174

Looks like we already added a deprecation warning for this.  It will be disallowed in airflow 3.0.  ~~Where should that warning have shown up?~~  I don't see a milestone tag on that PR, so I don't think the warning was in the version that @rustikk was testing.
